packages/worker/src/tasks/syncPackageVulnerabilities.ts

```diff
@@ -0,0 +1,140 @@
+} from '@gradejs-public/shared';
+
+type PackageVulnerabilityData = Pick<
+  PackageVulnerability,
+  'packageName' | 'packageVersionRange' | 'osvId' | 'osvData'
+>;
+
+async function drainStreamToBuffer(stream: Stream.Readable): Promise<Buffer> {
+  return new Promise((resolve, reject) => {
+    const dataChunks: any[] = [];
+
+    stream.on('data', (chunk) => dataChunks.push(chunk));
+    stream.on('end', () => resolve(Buffer.concat(dataChunks)));
+    stream.on('error', (err) => reject(err));
+
+    stream.resume();
+  });
+}
+
+function createPackageVulnerabilitySynchronizer(batchSize = 10) {
+  let entityBuffer: PackageVulnerabilityData[] = [];
+
+  const flush = async () => {
+    await PackageVulnerability.upsert(entityBuffer, {
+      skipUpdateIfNoValuesChanged: true,
+      conflictPaths: ['packageName', 'osvId'],
+    });
+    entityBuffer = [];
+  };
+
+  return async (osvData: OpenSourceVulnerability | null) => {
+    if (osvData === null) {
+      await flush();
+      return;
+    }
+
+    const affectedEntries = osvData.affected?.filter((it) => it.package.ecosystem === 'npm');
+    if (!affectedEntries?.length) {
+      return;
+    }
+
+    const affectedRangesByPackage: Record<string, string[]> = {};
+
+    for (const affectedEntry of affectedEntries) {
+      const affectedRanges = affectedEntry.ranges?.filter(
+        (it) => it.type === OSVAffectedRangeType.ECOSYSTEM
+      );
+      if (!affectedRanges) {
+        continue;
+      }
+
+      const affectedRangeParts = [];
+      for (const range of affectedRanges) {
+        const introducedVer = range.events.find((it) => !!it.introduced);
+        const fixedVer = range.events.find((it) => !!it.fixed);
+
+        if (introducedVer && fixedVer) {
+          affectedRangeParts.push(`>=${introducedVer.introduced} <${fixedVer.fixed}`);
+        }
+
+        if (introducedVer && !fixedVer) {
+          affectedRangeParts.push(`>=${introducedVer.introduced}`);
+        }
+      }
+
+      const affectedVersions = affectedEntry.versions;
+      if (affectedVersions) {
+        affectedRangeParts.push(...affectedVersions);
+      }
+
+      if (!affectedRangeParts.length) {
+        continue;
+      }
+
+      if (!affectedRangesByPackage[affectedEntry.package.name]) {
+        affectedRangesByPackage[affectedEntry.package.name] = affectedRangeParts;
+      } else {
+        affectedRangesByPackage[affectedEntry.package.name].push(...affectedRangeParts);
+      }
+    }
+
+    entityBuffer = entityBuffer.concat(
+      Object.entries(affectedRangesByPackage).map(([packageName, affectedRanges]) => ({
+        packageName,
+        packageVersionRange: affectedRanges.join(', '),
+        osvId: osvData.id,
+        osvData: osvData,
+      }))
+    );
+
+    if (entityBuffer.length >= batchSize) {
+      await flush();
+    }
+  };
+}
+
+export async function syncPackageVulnerabilities() {
+  const packageVulnerabilitySynchronizer = createPackageVulnerabilitySynchronizer();
```

Comment:
The `packageVulnerabilitySynchronizer` name looks like a name for an object. Can we rename it to something like `sync...`?